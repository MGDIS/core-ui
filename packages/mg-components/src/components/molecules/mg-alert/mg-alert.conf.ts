export { variants, variantStyles, type VariantStyleType, type VariantType } from '../mg-message/mg-message.conf';
