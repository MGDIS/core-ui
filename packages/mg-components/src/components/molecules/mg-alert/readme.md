## Usage

When a trigger (such as a button) is activated, it can display an alert message. By default, activating the trigger erases any previous alert messages of the same type on the screen.

## Hide component

To hide the component, you must use the native `hidden` attribute.

<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                                                                                                  | Type                                           | Default     |
| -------------- | --------------- | ------------------------------------------------------------------------------------------------------------ | ---------------------------------------------- | ----------- |
| `delay`        | `delay`         | Add a delay to hide/close message when it passed Value is defined in seconds and must greater than 2 seconds | `number`                                       | `undefined` |
| `variant`      | `variant`       | Define variant                                                                                               | `"danger" \| "info" \| "success" \| "warning"` | `undefined` |
| `variantStyle` | `variant-style` | Define variant style                                                                                         | `"background" \| "bar-left"`                   | `undefined` |


## Events

| Event             | Description                         | Type                |
| ----------------- | ----------------------------------- | ------------------- |
| `component-close` | Emmited event when alert is closed  | `CustomEvent<void>` |
| `component-hide`  | Emited event when alert is hidden   | `CustomEvent<void>` |
| `component-show`  | Emited event when alert is diplayed | `CustomEvent<void>` |


## Slots

| Slot        | Description     |
| ----------- | --------------- |
|             | Alert content   |
| `"actions"` | Actions content |


## Dependencies

### Depends on

- [mg-message](../mg-message)
- [mg-button](../../atoms/mg-button)
- [mg-icon](../../atoms/mg-icon)

### Graph
```mermaid
graph TD;
  mg-alert --> mg-message
  mg-alert --> mg-button
  mg-alert --> mg-icon
  mg-message --> mg-card
  mg-message --> mg-icon
  mg-button --> mg-icon
  style mg-alert fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
