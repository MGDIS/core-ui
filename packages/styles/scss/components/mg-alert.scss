.mg-c-alert {
  &__close-button {
    position: absolute;
    top: 0;
    right: 0;
  }

  &.mg-c-alert--close-button {
    --mg-c-card-padding: 0 var(--mg-b-size-min-height) 0 0;
  }
}
