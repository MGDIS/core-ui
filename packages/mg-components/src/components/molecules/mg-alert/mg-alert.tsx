import { Component, Element, Event, EventEmitter, h, Prop, State, Watch, forceUpdate } from '@stencil/core';
import { ClassList, toString } from '@mgdis/core-ui-helpers/stencil';
import { variants, variantStyles, VariantStyleType, VariantType } from './mg-alert.conf';
import { initLocales } from '../../../locales';

/**
 * @slot - Alert content
 * @slot actions - Actions content
 */
@Component({
  tag: 'mg-alert',
  styleUrl: '../../../../node_modules/@mgdis/styles/dist/components/mg-alert.css',
  shadow: true,
})
export class MgAlert {
  /************
   * Internal *
   ************/

  // Classes
  private readonly classCloseButton = 'mg-c-alert--close-button';

  // Stored timer setted when hide action is run from setTimeOut
  private storedTimer: ReturnType<typeof setTimeout> = null;

  // Manage hover
  private isFocused = false;
  private isHovered = false;

  // Locales
  private messages;

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgAlertElement;

  /**
   * Add a delay to hide/close message when it passed
   * Value is defined in seconds and must greater than 2 seconds
   */
  @Prop() delay?: number;
  @Watch('delay')
  watchDelay(newValue: MgAlert['delay']): void {
    if (newValue !== undefined && newValue < 2) {
      throw new Error(`<mg-alert> prop "delay" must be greater than 2 seconds. Passed value: ${toString(newValue)}.`);
    }
  }

  /**
   * Define variant
   */
  @Prop() variant?: VariantType = 'info';
  @Watch('variant')
  watchVariant(newValue: MgAlert['variant']): void {
    if (newValue && !variants.includes(newValue)) {
      throw new Error(`<mg-alert> prop "variant" must be one of: ${variants.join(', ')}. Passed value: ${toString(newValue)}.`);
    }
  }

  /**
   * Define variant style
   */
  @Prop() variantStyle?: VariantStyleType = 'bar-left';
  @Watch('variantStyle')
  watchVariantStyle(newValue: MgAlert['variantStyle']): void {
    if (newValue && !variantStyles.includes(newValue)) {
      throw new Error(`<mg-alert> prop "variantStyle" must be one of: ${variantStyles.join(', ')}. Passed value: ${toString(newValue)}.`);
    }
  }

  /**
   * Watch hidden prop
   */
  // eslint-disable-next-line @stencil-community/no-unused-watch
  @Watch('hidden')
  watchHidden(newValue: boolean): void {
    if (typeof newValue === 'string' && newValue === '') {
      newValue = true;
    }
    if (newValue) {
      this.componentHide.emit();
      // Remove event Listener
      ['focusin', 'mouseenter'].forEach(event => {
        this.element.removeEventListener(event, this.timerEvents);
      });
      // Clear Timer
      this.clearTimer();
    } else {
      this.componentShow.emit();
      // If delay is set
      if (this.delay > 1) {
        // Start timer
        this.setTimer();
        // Stop timer when get focus or mouse enter
        ['focusin', 'mouseenter'].forEach(event => {
          this.element.addEventListener(event, this.timerEvents);
        });
      }
    }
    forceUpdate(this);
  }

  /**
   * Component classes
   */
  @State() classCollection: ClassList = new ClassList(['mg-c-alert']);

  /**
   * Define if component is using actions slot
   */
  @State() hasActions = false;

  /**
   * Emited event when alert is diplayed
   */
  @Event({ eventName: 'component-show' }) componentShow: EventEmitter<void>;

  /**
   * Emited event when alert is hidden
   */
  @Event({ eventName: 'component-hide' }) componentHide: EventEmitter<void>;

  /**
   * Emmited event when alert is closed
   */
  @Event({ eventName: 'component-close' }) componentClose: EventEmitter<void>;

  /**
   * Set timer
   */
  private setTimer = (): void => {
    this.storedTimer = setTimeout(() => (this.element.hidden = true), this.delay * 1000);
  };

  /**
   * Clear timer
   */
  private clearTimer = (): void => {
    clearTimeout(this.storedTimer);
  };

  /**
   * Event to add on element
   * @param event - event
   */
  private timerEvents = (event: MouseEvent | FocusEvent): void => {
    this.clearTimer();
    const isMouseEvent: boolean = event.type === 'mouseenter';
    // Needed to ensure we don't start the timer
    this[isMouseEvent ? 'isHovered' : 'isFocused'] = true;

    // Restart timer when loose focus AND mouse leave
    this.element.addEventListener(
      isMouseEvent ? 'mouseleave' : 'focusout',
      eventOut => {
        this[eventOut.type === 'mouseleave' ? 'isHovered' : 'isFocused'] = false;
        // Start timer if needed
        if (!this.isFocused && !this.isHovered) this.setTimer();
      },
      { once: true },
    );
  };

  /**
   * Handle close button
   */
  private handleClose = (): void => {
    this.element.hidden = true;
    this.componentClose.emit();
  };

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if component props are well configured on init
   */
  componentWillLoad(): void {
    // Get locales
    this.messages = initLocales(this.element).messages;
    // Validate
    this.watchVariant(this.variant);
    this.watchVariantStyle(this.variantStyle);
    this.watchDelay(this.delay);
    this.watchHidden(this.element.hidden);
    // Check if component has actions slot
    this.hasActions = this.element.querySelector('[slot="actions"]') !== null;
    if (!this.hasActions) {
      this.classCollection.add(this.classCloseButton);
    }
  }

  /**
   * Render
   * @returns HTML Element
   */
  render(): HTMLElement {
    return (
      <mg-message
        class={this.classCollection.join()}
        role={['info', 'success'].includes(this.variant) ? 'status' : 'alert'}
        variant={this.variant}
        variantStyle={this.variantStyle}
      >
        <slot></slot>
        {this.hasActions ? (
          <span slot="actions">
            <slot name="actions"></slot>
          </span>
        ) : (
          <span class="mg-c-alert__close-button">
            <mg-button is-icon variant="flat" label={this.messages.alert.closeButton} onClick={this.handleClose}>
              <mg-icon icon="cross"></mg-icon>
            </mg-button>
          </span>
        )}
      </mg-message>
    );
  }
}
