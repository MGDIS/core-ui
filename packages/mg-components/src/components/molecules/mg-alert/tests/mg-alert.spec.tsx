import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgAlert } from '../mg-alert';
import { variants } from '../mg-alert.conf';

const getDefaultContent = () => (
  <p>
    <strong>Strong</strong> content!
  </p>
);

const getPage = (args, content) =>
  newSpecPage({
    components: [MgAlert],
    template: () => <mg-alert {...args}>{content}</mg-alert>,
  });

describe('mg-alert', () => {
  beforeEach(() => jest.useFakeTimers({ legacyFakeTimers: true }));
  afterEach(() => jest.runOnlyPendingTimers());

  describe('Render', () => {
    describe.each(variants)('Should render a %s alert', variant => {
      test.each([{}, { lang: 'fr' }, { lang: 'xx' }])('with args %s', async args => {
        const { root } = await getPage({ ...args, variant }, getDefaultContent());
        expect(root).toMatchSnapshot();
      });
    });

    test('Should render with action slot', async () => {
      const { root } = await getPage({}, [
        getDefaultContent(),
        <div slot="actions">
          <mg-button>Primary</mg-button>
          <mg-button variant="secondary">Secondary</mg-button>
        </div>,
      ]);
      expect(root).toMatchSnapshot();
    });
  });

  test('Should hide alert on close button', async () => {
    const page = await getPage({}, getDefaultContent());

    const element = page.doc.querySelector('mg-alert');
    const button = element.shadowRoot.querySelector('mg-button');

    jest.spyOn(page.rootInstance.componentHide, 'emit');

    button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
    await page.waitForChanges();

    expect(page.rootInstance.componentHide.emit).toHaveBeenCalledTimes(1);
    expect(page.rootInstance.element.hidden).toBe(true);
  });

  test('Should hide alert on delay', async () => {
    const args = { delay: 2 };
    const page = await getPage(args, getDefaultContent());

    jest.spyOn(page.rootInstance.componentHide, 'emit');
    jest.spyOn(page.rootInstance.componentShow, 'emit');

    expect(page.rootInstance.element.hidden).toBe(false);

    jest.advanceTimersByTime(args.delay * 1000);

    expect(page.rootInstance.element.hidden).toBe(true);
    expect(page.rootInstance.componentHide.emit).toHaveBeenCalledTimes(1);

    page.rootInstance.element.hidden = false;
    await page.waitForChanges();

    expect(page.rootInstance.componentShow.emit).toHaveBeenCalledTimes(1);

    jest.advanceTimersByTime(args.delay * 1000);

    expect(page.rootInstance.element.hidden).toBe(true);
    expect(page.rootInstance.componentHide.emit).toHaveBeenCalledTimes(2);
  });

  test.each(['focus', 'mouse'])('Should manage event %s with delay', async eventType => {
    const args = { delay: 2 };
    const page = await getPage(args, getDefaultContent());

    jest.spyOn(page.rootInstance.componentHide, 'emit');
    jest.spyOn(page.rootInstance.componentShow, 'emit');

    const element = page.doc.querySelector('mg-alert');

    expect(page.rootInstance.element.hidden).toBe(false);

    jest.advanceTimersByTime((args.delay / 2) * 1000);

    element.dispatchEvent(eventType === 'focus' ? new FocusEvent('focusin') : new MouseEvent('mouseenter'));

    jest.advanceTimersByTime(args.delay * 1000);

    expect(page.rootInstance.element.hidden).toBe(false);

    element.dispatchEvent(eventType === 'focus' ? new FocusEvent('focusout') : new MouseEvent('mouseleave'));

    jest.advanceTimersByTime(args.delay * 1000);

    expect(page.rootInstance.element.hidden).toBe(true);
    expect(page.rootInstance.componentHide.emit).toHaveBeenCalledTimes(1);
  });
  describe('Errors', () => {
    test.each([' ', 'blu'])('Should throw error with invalid variant property: %s', async variant => {
      expect.assertions(1);
      try {
        await getPage({ variant }, getDefaultContent());
      } catch (err) {
        expect(err.message).toMatch('<mg-alert> prop "variant" must be one of:');
      }
    });

    test.each([' ', 'batman'])('should throw error with invalid variantStyle property: %s', async variantStyle => {
      expect.assertions(1);
      try {
        await getPage(
          { variantStyle },
          'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
        );
      } catch (err) {
        expect(err.message).toMatch('<mg-alert> prop "variantStyle" must be one of:');
      }
    });

    test('Should throw error with invalid delay property', async () => {
      expect.assertions(1);
      try {
        await getPage({ delay: 1 }, getDefaultContent());
      } catch (err) {
        expect(err.message).toMatch('<mg-alert> prop "delay" must be greater than 2 seconds.');
      }
    });
  });
});
