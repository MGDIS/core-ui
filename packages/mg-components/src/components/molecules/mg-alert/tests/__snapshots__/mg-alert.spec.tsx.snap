// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-alert Render Should render a danger alert with args { closeButton: true } 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="danger">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Close message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a danger alert with args { closeButton: true, lang: 'fr' } 1`] = `
<mg-alert lang="fr">
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="danger">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Fermer le message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a danger alert with args { closeButton: true, lang: 'xx' } 1`] = `
<mg-alert lang="xx">
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="danger">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Close message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a danger alert with args {} 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert" role="alert" variant="danger">
      <slot></slot>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a info alert with args { closeButton: true } 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="status" variant="info">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Close message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a info alert with args { closeButton: true, lang: 'fr' } 1`] = `
<mg-alert lang="fr">
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="status" variant="info">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Fermer le message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a info alert with args { closeButton: true, lang: 'xx' } 1`] = `
<mg-alert lang="xx">
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="status" variant="info">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Close message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a info alert with args {} 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert" role="status" variant="info">
      <slot></slot>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a success alert with args { closeButton: true } 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="success">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Close message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a success alert with args { closeButton: true, lang: 'fr' } 1`] = `
<mg-alert lang="fr">
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="success">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Fermer le message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a success alert with args { closeButton: true, lang: 'xx' } 1`] = `
<mg-alert lang="xx">
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="success">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Close message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a success alert with args {} 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert" role="alert" variant="success">
      <slot></slot>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a warning alert with args { closeButton: true } 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="warning">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Close message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a warning alert with args { closeButton: true, lang: 'fr' } 1`] = `
<mg-alert lang="fr">
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="warning">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Fermer le message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a warning alert with args { closeButton: true, lang: 'xx' } 1`] = `
<mg-alert lang="xx">
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert mg-c-alert--close-button" role="alert" variant="warning">
      <slot></slot>
      <span class="mg-c-alert__close-button">
        <mg-button is-icon="" label="Close message" variant="flat">
          <mg-icon icon="cross"></mg-icon>
        </mg-button>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render a warning alert with args {} 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert" role="alert" variant="warning">
      <slot></slot>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
</mg-alert>
`;

exports[`mg-alert Render Should render with action slot 1`] = `
<mg-alert>
  <template shadowrootmode="open">
    <mg-message class="mg-c-alert" role="alert">
      <slot></slot>
      <span slot="actions">
        <slot name="actions"></slot>
      </span>
    </mg-message>
  </template>
  <p>
    <strong>
      Strong
    </strong>
    content!
  </p>
  <div slot="actions">
    <mg-button>
      Primary
    </mg-button>
    <mg-button variant="secondary">
      Secondary
    </mg-button>
  </div>
</mg-alert>
`;
